### ISSUE TITLE
Strange reward collapse during training

### ISSUE THREAD

**Issue by faustomilletari** [PrimeIntellect-ai/prime-rl#964]
Hello,

I wanted to describe an issue I'm encountering during training with verifiers + prime_rl.

If I train Qwen2.5-3B-Instruct with my own blend of different rewards using hugging face TRL and GRPOTrainer, the avg reward increases over time and eventually becomes very close to 1.0 (max reward) with a decreasing std. Training works.

If I train the same model and the same blend of rewards with verifiers and prime-rl, I often observe this situation. 

<img width="1055" height="526" alt="Image" src="https://github.com/user-attachments/assets/73cce96b-5f69-4431-a773-4a8b65f5a5ce" />

Initially I thought it was due to tool usage. But now, I have removed all tools and I'm doing the same thing I used to do with TRL (in essence). The graphs you see here do not use any tools, yet the avg reward collapses after a few thousand steps.

1 Node here means 8 GPUs training + 8 GPUs inference on two different machines (1 training node + 1 inference node)
4 Nodes here means 32 GPUs training + 8 GPUs inference on 5 different machines (4 training nodes + 1 inference node)

Currently it seems that the training using NCCL doesn't experience this issue, but it's too early to say.

Any pointer towards resolving or at least better understanding what's happening is greatly appreciated. Thank you!

**Comment by faustomilletari** [comment]
so far the NCCL-enabled job (1 node training + 1 node inference) didn't incur in that issue. (same dataset, same configs).

**Comment by samsja** [comment]
> Hello,
> 
> I wanted to describe an issue I'm encountering during training with verifiers + prime_rl.
> 
> If I train Qwen2.5-3B-Instruct with my own blend of different rewards using hugging face TRL and GRPOTrainer, the avg reward increases over time and eventually becomes very close to 1.0 (max reward) with a decreasing std. Training works.
> 
> If I train the same model and the same blend of rewards with verifiers and prime-rl, I often observe this situation.
> 
> <img alt="Image" width="1055" height="526" src="https://private-user-images.githubusercontent.com/13331646/491700752-73cce96b-5f69-4431-a773-4a8b65f5a5ce.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgzMTM4MDYsIm5iZiI6MTc1ODMxMzUwNiwicGF0aCI6Ii8xMzMzMTY0Ni80OTE3MDA3NTItNzNjY2U5NmItNWY2OS00NDMxLWE3NzMtNGE4YjY1ZjVhNWNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTE5VDIwMjUwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU4MzQ2NmQ5ZjQ4YWQwZTM3MTMyZmViMGE1MTkwNDlmMmJhZGNjMTRiZWFkMDFjODkzMTZjYmRiNmE1NjVkNzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tbU1ZeRVbf1RFEgKlg7JDrbUEzNgGaI3-ZI6Gxf81Bw">
> Initially I thought it was due to tool usage. But now, I have removed all tools and I'm doing the same thing I used to do with TRL (in essence). The graphs you see here do not use any tools, yet the avg reward collapses after a few thousand steps.
> 
> 1 Node here means 8 GPUs training + 8 GPUs inference on two different machines (1 training node + 1 inference node) 4 Nodes here means 32 GPUs training + 8 GPUs inference on 5 different machines (4 training nodes + 1 inference node)
> 
> Currently it seems that the training using NCCL doesn't experience this issue, but it's too early to say.
> 
> Any pointer towards resolving or at least better understanding what's happening is greatly appreciated. Thank you!

Hey thank you for reporting this issue:

Couple of question:

* can you also share the screenshot for grad norm and entropy loss ?
* can you share the set if hyper parameters you are using (batch size, lr, number of async step)
* I am not sure to get what is the difference between the NCCL group and 4 nodes ? Also seems weird that we observe a different training dynamic with multi node, any detail on the difference setup would be help us investigating

**Comment by samsja** [comment]
Also do you have a run with the x axis being number of steps rather that time ? I find it curious that the multi node one crash after the same amount of time as the single node, I would expect the 4 nodes to crash at the same step count so roughly a quarter of the time of the 1 node

**Comment by faustomilletari** [comment]
Hello,

here the hyperparameters:

**orch.toml**
max_steps = 100000
batch_size = 256
micro_batch_size = 1
seq_len = 4096
rollouts_per_example = 16
mask_truncated_completions = true

...

**train.toml**
max_steps = 100000

[wandb]
project='rbio'

[model]
name = "Qwen/Qwen2.5-3B-Instruct"

[optim]
lr = 1e-6
max_norm = 0.5

[loss]
clip_ratio = 0.7

[ckpt]
interval = 1000
keep = 1

async steps is the default value.

grad norm and entropy

<img width="1063" height="519" alt="Image" src="https://github.com/user-attachments/assets/83421f92-ad6b-41e7-aca7-8d6880f3fb18" />

<img width="1092" height="630" alt="Image" src="https://github.com/user-attachments/assets/4f162829-7fbc-47c0-b57d-359d9c9e4f94" />

reward avg vs. steps

<img width="1092" height="526" alt="Image" src="https://github.com/user-attachments/assets/507e1db4-203d-4a7c-b71e-93cbf8d40cb6" />

The configs and everything really of 1 Node (plus 1 inference node) (8 GPUs training, 8 GPUS inference) with and without NCCL are the same. I had used env variables to disable NCCL in the run without NCCL.

**Comment by samsja** [comment]
So Node NCCL is with or without NCCL? Can you precise which environment variable you are using ?

**Comment by faustomilletari** [comment]
1 Node NCCL = 1 training node **with** NCCL + 1 inference node also with NCCL
1 Node = 1 training node **without** NCCL + 1 inference node also without NCCL

```
export NCCL_P2P_DISABLE=1
export NCCL_IB_DISABLE=1
export NCCL_SHM_DISABLE=1
export NCCL_NET_GDR_LEVEL=0
export NCCL_DEBUG=INFO
```

thank you so much for your help and all the great work here!

**Comment by samsja** [comment]
> [loss]
> clip_ratio = 0.7

The clip ratio looks off, it should be a value above 1, could you try to run with the default value 8 ?

**Comment by samsja** [comment]
> 1 Node NCCL = 1 training node **with** NCCL + 1 inference node also with NCCL 1 Node = 1 training node **without** NCCL + 1 inference node also without NCCL
> 
> ```
> export NCCL_P2P_DISABLE=1
> export NCCL_IB_DISABLE=1
> export NCCL_SHM_DISABLE=1
> export NCCL_NET_GDR_LEVEL=0
> export NCCL_DEBUG=INFO
> ```
> 
> thank you so much for your help and all the great work here!

oh I see so NCCL is not really disable here, it just p2p so its still using nccl but going over the pcie instead of nvlink. I highly doubt this could have an impact on the stability of the training. Is there any other difference between them ?

**Comment by faustomilletari** [comment]
Ok, I will make the changes and create new runs (to completion) with and without those same env variables.

Thank you!

**Comment by samsja** [comment]
> > [loss]
> > clip_ratio = 0.7
> 
> The clip ratio looks off, it should be a value above 1, could you try to run with the default value 8 ?

if you can plot here the importance clipping graph we could already see if the 0.7 clip ratio has had a big impact on the stability

**Comment by samsja** [comment]
> Ok, I will make the changes and create new runs (to completion) with and without those same env variables.
> 
> Thank you!

you are welcome, happy to help :)

**Comment by faustomilletari** [comment]
This is still unsolved but I have decided to first focus on obtaining a sensible speedup via multi-node training.

**Comment by mikasenghaas** [comment]
we're very interested in resolving this i think, would be help a ton if you could share the exact training setup used in trl and prime-rl to be able to circle out possible differences. would guess that it has smth to do with async rl, i.e. our loss objective, importance ratios, distribution shift between trainer/ infer etc. also we have no kl by default which might cause instability as training saturates. many possible reasons, but need some more details to really say:)

**Comment by faresobeid** [comment]
Will close for now bcs of staleness, feel free to re-open
